### PR TITLE
Loader using Module .cctor

### PR DESCRIPTION
**Pull request by tonyredondo**
**This replaces #1075, it was simpler recreate the PR than rebasing it.**

--- 

This PR contains:

- [x] Refactor of the loader logic to a new `loader.h/loader.cpp` file.
- [x] Change the implementation to use `<Module>..ctor` to call the startup hook.
- [x] Include a lock at the native level to avoid loading the loader multiple times in the same AppDomainID.
- [x] Skip the default domain in IIS 

Example:
```log
12/10/20 06:55:42.559 PM [dotnet.exe|3256|11440] [info] Datadog CLR Profiler 1.21.0 64-bit on Windows
12/10/20 06:55:42.560 PM [dotnet.exe|3256|11440] [info] Environment variables:
12/10/20 06:55:42.560 PM [dotnet.exe|3256|11440] [info]   DD_TRACE_DEBUG=0
12/10/20 06:55:42.560 PM [dotnet.exe|3256|11440] [info]   DD_DOTNET_TRACER_HOME=C:\github\dd-trace-dotnet\src\bin\tracer-home
12/10/20 06:55:42.560 PM [dotnet.exe|3256|11440] [info]   DD_INTEGRATIONS=C:\github\dd-trace-dotnet\src\bin\tracer-home\integrations.json
12/10/20 06:55:42.560 PM [dotnet.exe|3256|11440] [info]   DD_ENV=tony
12/10/20 06:55:42.560 PM [dotnet.exe|3256|11440] [info]   DD_DUMP_ILREWRITE_ENABLED=0
12/10/20 06:55:42.568 PM [dotnet.exe|3256|11440] [info] Profiler attached.
12/10/20 06:55:42.660 PM [dotnet.exe|3256|11440] [info] ModuleLoadFinished: Datadog.Trace.ClrProfiler.Managed.Loader loaded into AppDomain 2429526899488 clrhost

12/10/20 06:55:42.702 PM [dotnet.exe|3256|11440] [warning] Loader::GetAssemblyAndSymbolsBytes the loader was already loaded for AppDomainID=2429526899488

12/10/20 06:55:42.778 PM [dotnet.exe|3256|11440] [info] AssemblyLoadFinished: Datadog.Trace.ClrProfiler.Managed v1.21.0 matched profiler version v1.21.0
```

### Usage

To use it we create the loader class in the `Initialize` method:

```
  // Create the loader class
  loader_ = new Loader(this->info_, false);
```

Then in the ModuleLoadFinished callback:

```
  // Inject loader to the module initializer
  loader->InjectLoaderToModuleInitializer(module_id);
```

### CODE REVIEW

**Review by macrogreg — CHANGES_REQUESTED** [review]

File: src/Datadog.Trace.ClrProfiler.Native/loader.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        ICorProfilerInfo4* info_;
+        bool is_iis_;
+
+        std::mutex loaders_loaded_mutex_;
+        std::unordered_set<AppDomainID> loaders_loaded_;
+
+    public:
+        Loader(ICorProfilerInfo4* info, bool isIIS);
+
+        HRESULT InjectLoaderToModuleInitializer(const ModuleID module_id);
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.h:23]: This is called from [CorProfiler::ModuleLoadFinished](https://github.com/DataDog/dd-trace-dotnet/blob/tony/loader-with-modules-cctor/src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp#L386-L387) callback.

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 2]
```diff
@@ -0,0 +1,909 @@
...
+
+#ifdef BIT64
+        WSTRING native_profiler_file = GetEnvironmentValue("CORECLR_PROFILER_PATH_64"_W);
+        Debug("Loader::InjectLoaderToModuleInitializer: CORECLR_PROFILER_PATH_64 defined as: ", native_profiler_file);
+
+        if (native_profiler_file == ""_W) {
+            native_profiler_file = GetEnvironmentValue("CORECLR_PROFILER_PATH"_W);
+            Debug("Loader::InjectLoaderToModuleInitializer: CORECLR_PROFILER_PATH defined as: ", native_profiler_file);
+        }
+#else   // BIT64
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp:280]: yeah, we support windows x86

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 3]
```diff
@@ -0,0 +1,909 @@
...
+            return hr;
+        }
+
+        //
+        // create assembly ref to mscorlib
+        //
+        mdAssemblyRef mscorlib_ref = mdAssemblyRefNil;
+        ASSEMBLYMETADATA metadata{ 4, 0, 0, 0};
+        BYTE public_key[] = {0xB7, 0x7A, 0x5C, 0x56, 0x19, 0x34, 0xE0, 0x89};
+        hr = assembly_emit->DefineAssemblyRef(public_key, sizeof(public_key), "mscorlib"_W.c_str(), &metadata, NULL, 0, 0, &mscorlib_ref);
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp:312]: yes, I just move this code from the current profiler loader.

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 4]
```diff
@@ -0,0 +1,909 @@
...
+        if (FAILED(hr)) {
+            Warn("Loader::InjectLoaderToModuleInitializer: Error creating assembly reference to mscorlib.");
+            return hr;
+        }
+
+        //
+        // get a TypeRef for System.Byte
+        //
+        mdTypeRef byte_type_ref;
+        hr = metadata_emit->DefineTypeRefByName(mscorlib_ref, "System.Byte"_W.c_str(), &byte_type_ref);
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp:322]: no, I didn't have this problem in my tests. This method should return the same mdToken that is already declared.

https://github.com/dotnet/runtime/blob/69e114c1abf91241a0eeecf1ecceab4711b8aa62/src/coreclr/md/compiler/regmeta_emit.cpp#L1015

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 5]
```diff
@@ -0,0 +1,909 @@
...
+                &assembly_create_instance_member_ref);
+        if (FAILED(hr)) {
+            Warn("Loader::InjectLoaderToModuleInitializer: DefineMemberRef failed");
+            return hr;
+        }
+
+        // Create a string representing
+        // "Datadog.Trace.ClrProfiler.Managed.Loader.Startup" Create OS-specific
+        // implementations because on Windows, creating the string via
+        // "Datadog.Trace.ClrProfiler.Managed.Loader.Startup"_W.c_str() does not
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp:479]: This is copied from the existing loader, I have to recheck all of these conditions, I want to remove the maximum number of `_W` operator calls.

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 6]
```diff
@@ -0,0 +1,909 @@
...
+        HGLOBAL hResSymbols = LoadResource(hInstance, hResSymbolsInfo);
+        *symbolsSize = SizeofResource(hInstance, hResSymbolsInfo);
+        *pSymbolsArray = (LPBYTE)LockResource(hResSymbols);
+#elif LINUX
+        *assemblySize = dll_end - dll_start;
+        *pAssemblyArray = (BYTE*)dll_start;
+
+        *symbolsSize = pdb_end - pdb_start;
+        *pSymbolsArray = (BYTE*)pdb_start;
+#else
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp]: done.

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 7]
```diff
@@ -0,0 +1,909 @@
...
+        const unsigned int imgCount = _dyld_image_count();
+
+        for(auto i = 0; i < imgCount; i++) {
+            const std::string name = std::string(_dyld_get_image_name(i));
+
+            if (name.rfind("Datadog.Trace.ClrProfiler.Native.dylib") != std::string::npos) {
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp]: All string literals are class-level const.

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 8]
```diff
@@ -0,0 +1,909 @@
...
+            return hr;
+        }
+
+        //
+        // retrieve AppDomainID from AssemblyID
+        //
+        AppDomainID app_domain_id = 0;
+        WCHAR assembly_name[250];
+        DWORD assembly_name_len = 0;
+        hr = this->info_->GetAssemblyInfo(assembly_id, 100, &assembly_name_len,
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp]: fixed.

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 9]
```diff
@@ -0,0 +1,909 @@
...
+        auto assembly_name_string = WSTRING(assembly_name);
+
+        //
+        // retrieve AppDomain Name
+        //
+
+        WCHAR app_domain_name[250];
+        DWORD app_domain_name_len = 0;
+
+        hr = this->info_->GetAppDomainInfo(app_domain_id, kNameMaxSize,
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp]: done.

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 10]
```diff
@@ -0,0 +1,909 @@
...
+                                    &app_domain_name_len, app_domain_name,
+                                    nullptr);
+
+        WSTRING app_domain_name_string;
+        if (SUCCEEDED(hr)) {
+          app_domain_name_string = WSTRING(app_domain_name);
+        } else {
+          app_domain_name_string = ""_W;
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp]: I was able to remove all `_W` to `L` in newer commits.

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 11]
```diff
@@ -0,0 +1,909 @@
...
+        // If we are in the IIS process we skip the default domain
+        if (is_iis_ && app_domain_name_string == "DefaultDomain"_W) {
+            Debug("Loader::InjectLoaderToModuleInitializer: Skipping ", assembly_name_string, ". The module belongs to the DefaultDomain in IIS process.");
+            return E_FAIL;
+        }
+
+        //
+        // check if the module is not the loader itself
+        //
+        if (assembly_name_string == "Datadog.Trace.ClrProfiler.Managed.Loader"_W) {
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp]: All string literals are class-level const.

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 12]
```diff
@@ -0,0 +1,909 @@
...
+        }
+
+        metadata_emit->SetMethodImplFlags(pinvoke_method_def, miPreserveSig);
+        if (FAILED(hr)) {
+            Warn("Loader::InjectLoaderToModuleInitializer: SetMethodImplFlags for GetAssemblyAndSymbolsBytes failed.");
+            return hr;
+        }
+
+#ifdef _WIN32
+        WSTRING native_profiler_file = "DATADOG.TRACE.CLRPROFILER.NATIVE.DLL"_W;
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp]: All string literals are class-level const now.

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp]: I was able to remove all _W to L in newer commits.

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 13]
```diff
@@ -0,0 +1,909 @@
...
+
+        // Create a string representing
+        // "Datadog.Trace.ClrProfiler.Managed.Loader.Startup" Create OS-specific
+        // implementations because on Windows, creating the string via
+        // "Datadog.Trace.ClrProfiler.Managed.Loader.Startup"_W.c_str() does not
+        // create the proper string for CreateInstance to successfully call
+#ifdef _WIN32
+        LPCWSTR load_helper_str = L"Datadog.Trace.ClrProfiler.Managed.Loader.Startup";
+        auto load_helper_str_size = wcslen(load_helper_str);
+#else
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp]: done.

File: src/Datadog.Trace.ClrProfiler.Native/loader.cpp [hunk 14]
```diff
@@ -0,0 +1,909 @@
...
+        BYTE initialize_signature[] = {
+                IMAGE_CEE_CS_CALLCONV_DEFAULT,  // Calling convention
+                0,                              // Number of parameters
+                ELEMENT_TYPE_VOID,              // Return type
+                ELEMENT_TYPE_OBJECT             // List of parameter types
+        };
+
+        mdMethodDef startup_method_def;
+        hr = metadata_emit->DefineMethod(
+                module_type_def, "__DDVoidMethodCall__"_W.c_str(), mdStatic,
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/loader.cpp]: done in 4e94773

**Comment by tonyredondo** [comment]
This will be handled in the future with a new native loader.

**Pull request closed without merging**